### PR TITLE
doc: capture device-plugin stricter workflow ordering explicitly

### PR DESCRIPTION
**Pull request by swatisehgal**
Based on kubelet device manager refactoring done in 1.25 release, there is stricter ordering requirements where the device plugin MUST start a gRPC service before registering itself to kubelet. In case this ordering is not followed, the plugin registration will fail.

Signed-off-by: Swati Sehgal <swsehgal@redhat.com>

Fixes https://github.com/kubernetes/kubernetes/issues/112395

### CODE REVIEW

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 1]
```diff
@@ -132,6 +132,11 @@ The general workflow of a device plugin includes the following steps:
 * The plugin registers itself with the kubelet through the Unix socket at host
   path `/var/lib/kubelet/device-plugins/kubelet.sock`.
 
+  {{< note >}}
```
Comment (mythi) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:135]: I'm thinking we could also move the workflow to an ordered list?

Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:135]: Sure, done.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 2]
```diff
@@ -129,10 +129,15 @@ The general workflow of a device plugin includes the following steps:
...
   {{< /note >}}
 
-* The plugin registers itself with the kubelet through the Unix socket at host
+3. The plugin registers itself with the kubelet through the Unix socket at host
   path `/var/lib/kubelet/device-plugins/kubelet.sock`.
 
-* After successfully registering itself, the device plugin runs in serving mode, during which it keeps
+  {{< note >}}
+  The ordering of the workflow is important. A plugin MUST start serving gRPC
+  server before registering itself with kubelet for successfull registeration.
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Fixed, Thanks.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 3]
```diff
@@ -129,10 +129,15 @@ The general workflow of a device plugin includes the following steps:
   available, before calling any of them directly.
   {{< /note >}}
 
-* The plugin registers itself with the kubelet through the Unix socket at host
+3. The plugin registers itself with the kubelet through the Unix socket at host
   path `/var/lib/kubelet/device-plugins/kubelet.sock`.
 
-* After successfully registering itself, the device plugin runs in serving mode, during which it keeps
+  {{< note >}}
+  The ordering of the workflow is important. A plugin MUST start serving gRPC
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:136]: Yes, that's right. Fixed it. Thanks for identifying this.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 4]
```diff
@@ -85,10 +85,10 @@ spec:
 
 The general workflow of a device plugin includes the following steps:
 
-* Initialization. During this phase, the device plugin performs vendor specific
+1. Initialization. During this phase, the device plugin performs vendor specific
   initialization and setup to make sure the devices are in a ready state.
 
-* The plugin starts a gRPC service, with a Unix socket under host path
+2. The plugin starts a gRPC service, with a Unix socket under host path
```
Comment (tengqm) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:91]: Even for an ordered list, we don't explicit use integers other than `1`. The Markdown parser will figure out the correct number to use.
The benefit of always using `1` is that we can always insert/remove the items without having to renumber the whole list again.

Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: updated it, Thanks.

**Comment by mehabhalodiya** [comment]
👋🏻 @swatisehgal 
We couldn't merge this as-is as there is a conflict that the automatic system does not know how to resolve.

Here are some references/tips for the following:

- [merge conflicts and rebasing](https://kubernetes.io/docs/contribute/new-content/open-a-pr/#merge-conflicts-and-rebasing)
- [squashing commits](https://kubernetes.io/docs/contribute/new-content/open-a-pr/#squashing-commits)

Another option would be:
Close this pull request, and open a new one based on the main branch that makes the same changes. In the new PR, mention that it's a replacement for PR https://github.com/kubernetes/website/pull/36940

Thank you.

**Comment by tengqm** [comment]
@swatisehgal Please try resolve the conflicts? This one has been sitting there for too long.

**Comment by swatisehgal** [comment]
Thanks @mehabhalodiya and @tengqm for the ping. I have rebased the PR.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 5]
```diff
@@ -87,60 +87,65 @@ spec:
...
+    service before registering itself with kubelet for successful registration.
+    {{< /note >}}
+
+1. After successfully registering itself, the device plugin runs in serving mode, during which it keeps
+monitoring device health and reports back to the kubelet upon any device state changes.
+It is also responsible for serving `Allocate` gRPC requests. During `Allocate`, the device plugin may
+do device-specific preparation; for example, GPU cleanup or QRNG initialization.
+If the operations succeed, the device plugin returns an `AllocateResponse` that contains container
+runtime configurations for accessing the allocated devices. The kubelet passes this information
+to the container runtime.
```
Comment (tengqm) [inline content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md:148]: I'd suggest we indent the who paragraph for readability.

```suggestion
   monitoring device health and reports back to the kubelet upon any device state changes.
   It is also responsible for serving `Allocate` gRPC requests. During `Allocate`, the device plugin may
   do device-specific preparation; for example, GPU cleanup or QRNG initialization.
   If the operations succeed, the device plugin returns an `AllocateResponse` that contains container
   runtime configurations for accessing the allocated devices. The kubelet passes this information
   to the container runtime.
```

Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Done, Thanks!

**Review by divya-mohan0209 — COMMENTED** [review]
non-blocking nits. Overall LGTM otherwise!

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 6]
```diff
@@ -87,60 +87,65 @@ spec:
 
 The general workflow of a device plugin includes the following steps:
 
-* Initialization. During this phase, the device plugin performs vendor specific
+1. Initialization. During this phase, the device plugin performs vendor specific
   initialization and setup to make sure the devices are in a ready state.
 
-* The plugin starts a gRPC service, with a Unix socket under host path
+1. The plugin starts a gRPC service, with a Unix socket under host path
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Thanks, fixed.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 7]
```diff
@@ -87,60 +87,65 @@ spec:
...
+          // PreStartContainer is called, if indicated by Device Plugin during registeration phase,
+          // before each container start. Device plugin can run device specific operations
+          // such as resetting the device before making devices available to the container.
+          rpc PreStartContainer(PreStartContainerRequest) returns (PreStartContainerResponse) {}
+    }
+    ```
+
+    {{< note >}}
+    Plugins are not required to provide useful implementations for
+    `GetPreferredAllocation()` or `PreStartContainer()`. Flags indicating which
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Thanks, fixed.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 8]
```diff
@@ -87,60 +87,65 @@ spec:
...
+    (if any) of these calls are available should be set in the `DevicePluginOptions`
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Thanks, fixed.

File: content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md [hunk 9]
```diff
@@ -87,60 +87,65 @@ spec:
 
 The general workflow of a device plugin includes the following steps:
 
-* Initialization. During this phase, the device plugin performs vendor specific
+1. Initialization. During this phase, the device plugin performs vendor specific
```
[the remark this answers is not in the document]
Comment (swatisehgal) [reply content/en/docs/concepts/extend-kubernetes/compute-storage-net/device-plugins.md]: Thanks, fixed.

**Review by divya-mohan0209 — COMMENTED** [review]
LGTM from a docs perspective! 

Do we have anyone who can peer review/provide signoff from a tech perspective, @swatisehgal ? I can add my approval after they're done providing their signoff.

**Comment by mythi** [comment]
> Do we have anyone who can peer review/provide signoff from a tech perspective, @swatisehgal ? I can add my approval after they're done providing their signoff.

FWIW, I've LGTM'd it as a contributor to the original issue (https://github.com/kubernetes/kubernetes/issues/112395#issuecomment-1253631034) and as a maintainer of a vendor device plugin.

**Comment by swatisehgal** [comment]
> FWIW, I've LGTM'd it as a contributor to the original issue ([kubernetes/kubernetes#112395 (comment)](https://github.com/kubernetes/kubernetes/issues/112395#issuecomment-1253631034)) and as a maintainer of a vendor device plugin.

@mythi I believe it is time to become an official Kubernetes member :)  https://github.com/kubernetes/community/blob/master/community-membership.md#member

**Review by divya-mohan0209 — COMMENTED** [review]
Thank you for the clarification @mythi 
/approve
/lgtm

**Pull request merged**